src/python/tests/gff3/test_restructure.py

```diff
@@ -0,0 +1,314 @@
+            feat = SeqFeature(loc, type=ftype)
+            feat.qualifiers["source"] = self.source
+            feat.sub_features = []
+            feats.append(feat)
+        return feats
+
+    def make_structure(self, children: List[Any]) -> List[SeqFeature]:
+        """Return a SeqFeature children structure from the form:
+        struct = ["mRNA"]
+        struct = [{"mRNA": ["CDS", "exon"]}, "exon", "exon"]
+        """
+        output = []
+        for child in children:
+            if isinstance(child, str):
+                output += self.make(child)
+            elif isinstance(child, dict):
+                feat_type = list(child.keys())[0]
+                feat_children = list(child.values())[0]
+
+                feat = self.make(feat_type)[0]
+                feat.sub_features += self.make_structure(feat_children)
+                output.append(feat)
+
+        return output
+
+    def get_sub_structure(self, feat: SeqFeature) -> Union[Dict, str]:
+        """Create a children structure from a SeqFeature."""
+        if feat.sub_features:
+            feat_subs = []
+            for sub in feat.sub_features:
+                feat_subs.append(self.get_sub_structure(sub))
+            return {feat.type: feat_subs}
+        return feat.type
+
+
+@pytest.mark.parametrize(
+    "children, expected_children",
+    [
+        param(["gene"], [{"gene": ["transcript"]}], id="gene, no transcript: add"),
+        param([{"gene": ["mRNA"]}], [{"gene": ["mRNA"]}], id="gene + mRNA"),
+        param([{"gene": ["mRNA", "mRNA"]}], [{"gene": ["mRNA", "mRNA"]}], id="gene + 2 mRNA"),
+        param(["ncRNA_gene"], ["ncRNA_gene"], id="1 ncRNA_gene, no transcript"),
+        param(
+            [{"ncRNA_gene": ["transcript"]}], [{"ncRNA_gene": ["transcript"]}], id="1 ncRNA_gene + transcript"
+        ),
+    ],
+)
+def test_add_transcript_to_naked_gene(children: List[Any], expected_children: List[Any]):
+    """Test the creation of a transcript for a gene without one."""
+    gen = FeatGenerator()
+    genes = gen.make_structure(children)
+    add_transcript_to_naked_gene(genes[0])
+    assert gen.get_sub_structure(genes[0]) == expected_children[0]
+
+
+@pytest.mark.parametrize(
+    "children, expected_children, expectation",
+    [
+        param(["mRNA"], ["mRNA"], does_not_raise(), id="One mRNA, skip"),
+        param(["CDS"], [{"mRNA": ["CDS", "exon"]}], does_not_raise(), id="One CDS, add mRNA"),
+        param(
+            ["CDS", "CDS"],
+            [{"mRNA": ["CDS", "exon", "CDS", "exon"]}],
+            does_not_raise(),
+            id="2 CDSs, add mRNA",
+        ),
+        param(["exon"], ["exon"], does_not_raise(), id="One exon, skip"),
+        param(["CDS", "exon"], ["CDS", "exon"], does_not_raise(), id="1 CDS + 1 exon, skip"),
+    ],
+)
+def test_move_only_cdss_to_new_mrna(
+    children: List[str],
+    expected_children: Dict[str, int],
+    expectation: ContextManager,
```

Comment:
This is only necessary if you expect at least one test case to raise an exception. Otherwise, I would suggest to remove it from the arguments, parametrisation and the `with expectation` below.